File: lib/models/tournament.dart
import 'player.dart';

class Tournament {
  const Tournament({
    required this.id,
    required this.name,
    required this.discipline,
    required this.location,
    required this.surface,
    required this.category,
    required this.format,
    required this.startDate,
    required this.endDate,
    required this.status,
    required this.visibility,
    required this.friendly,
    this.owner,
    this.admins = const [],
    required this.participants,
    this.images = const [],
    this.winner,
  });

  final String id;
  final String name;
  final String discipline;
  final String location;
  final String surface;
  final String category;
  final String format;
  final DateTime startDate;
  final DateTime endDate;
  final String status;
  final String visibility;
  final bool friendly;
  final Player? owner;
  final List<Player> admins;
  final List<Player> participants;
  final List<String> images;
  final Player? winner;

  bool get isUpcoming => status == 'upcoming';
  bool get isPrivate => visibility == 'private';

  bool canManage(String? playerId, {bool appAdmin = false}) {
    if (appAdmin) return true;
    if (playerId == null) return false;
    return owner?.id == playerId || admins.any((admin) => admin.id == playerId);
  }

  factory Tournament.fromJson(Map<String, dynamic> json) {
    return Tournament(
      id: (json['_id'] ?? json['id']).toString(),
      name: json['name'] ?? '',
      discipline: json['discipline'] ?? 'singles',
      location: json['location'] ?? '',
      surface: json['surface'] ?? 'Hard',
      category: json['category'] ?? '',
      format: json['format'] ?? 'elimination',
      startDate: DateTime.tryParse(json['startDate'] ?? '') ?? DateTime.now(),
      endDate: DateTime.tryParse(json['endDate'] ?? '') ?? DateTime.now(),
      status: json['status'] ?? 'upcoming',
      visibility: json['visibility'] ?? 'public',
      friendly: json['friendly'] == true,
      owner: json['owner'] is Map<String, dynamic>
          ? Player.fromJson(json['owner'])
          : null,
      admins: (json['admins'] as List? ?? [])
          .whereType<Map<String, dynamic>>()
          .map(Player.fromJson)
          .toList(),
      participants: (json['participants'] as List? ?? [])
          .whereType<Map<String, dynamic>>()
          .map(Player.fromJson)
          .toList(),
      images: (json['images'] as List? ?? [])
          .map((item) => item.toString())
          .toList(),
      winner: json['winner'] is Map<String, dynamic>
          ? Player.fromJson(json['winner'])
          : null,
    );
  }
}

class TournamentRanking {
  const TournamentRanking({
    required this.player,
    required this.wins,
    required this.losses,
    required this.matchesPlayed,
    required this.points,
  });

  final Player player;
  final int wins;
  final int losses;
  final int matchesPlayed;
  final int points;

  factory TournamentRanking.fromJson(Map<String, dynamic> json) {
    return TournamentRanking(
      player: Player.fromJson(json['player'] ?? <String, dynamic>{}),
      wins: json['wins'] ?? 0,
      losses: json['losses'] ?? 0,
      matchesPlayed: json['matchesPlayed'] ?? 0,
      points: json['points'] ?? 0,
    );
  }
}

File: backend/src/models/Tournament.ts
import { Schema, model, type Types } from "mongoose";

export type TournamentStatus = "upcoming" | "active" | "finished";
export type TournamentVisibility = "public" | "private";
export type TournamentFormat =
  | "elimination"
  | "round_robin"
  | "qualification"
  | "group_knockout"
  | "double_elimination"
  | "compass"
  | "swiss";

export interface TournamentAttrs {
  name: string;
  discipline: "singles" | "doubles";
  location: string;
  locations: Types.ObjectId[];
  surface: string;
  category: string;
  format: TournamentFormat;
  knockoutSize: number;
  knockoutSeeds: Types.ObjectId[];
  knockoutStartedAt?: Date;
  knockoutRounds: { round: string; matchIds: Types.ObjectId[] }[];
  startDate: Date;
  endDate: Date;
  status: TournamentStatus;
  visibility: TournamentVisibility;
  friendly: boolean;
  owner: Types.ObjectId;
  admins: Types.ObjectId[];
  participants: Types.ObjectId[];
  matches: Types.ObjectId[];
  images: string[];
  bracketSize?: number;
  drawGeneratedAt?: Date;
  winner?: Types.ObjectId;
  awardApplied?: boolean;
  awardOperationId?: Types.ObjectId;
  awardWinPoints?: number;
}

const tournamentSchema = new Schema<TournamentAttrs>(
  {
    name: { type: String, required: true, trim: true },
    discipline: { type: String, enum: ["singles", "doubles"], default: "singles" },
    location: { type: String, required: true, trim: true },
    locations: [{ type: Schema.Types.ObjectId, ref: "Location" }],
    surface: { type: String, required: true, trim: true, default: "Hard" },
    category: { type: String, required: true, trim: true },
    format: {
      type: String,
      enum: ["elimination", "round_robin", "qualification", "group_knockout", "double_elimination", "compass", "swiss"],
      default: "elimination"
    },
    startDate: { type: Date, required: true },
    knockoutSize: { type: Number, enum: [0, 2, 4, 8, 16], default: 0 },
    knockoutSeeds: [{ type: Schema.Types.ObjectId, ref: "Player" }],
    knockoutStartedAt: { type: Date },
    knockoutRounds: [{
      _id: false,
      round: { type: String, required: true },
      matchIds: [{ type: Schema.Types.ObjectId, ref: "Match" }]
    }],
    endDate: { type: Date, required: true },
    status: { type: String, enum: ["upcoming", "active", "finished"], default: "upcoming" },
    visibility: { type: String, enum: ["public", "private"], default: "public" },
    friendly: { type: Boolean, default: false },
    owner: { type: Schema.Types.ObjectId, ref: "Player", required: true },
    admins: [{ type: Schema.Types.ObjectId, ref: "Player" }],
    participants: [{ type: Schema.Types.ObjectId, ref: "Player" }],
    matches: [{ type: Schema.Types.ObjectId, ref: "Match" }],
    images: { type: [String], default: [] },
    bracketSize: { type: Number },
    drawGeneratedAt: { type: Date },
    winner: { type: Schema.Types.ObjectId, ref: "Player" },
    awardApplied: { type: Boolean, select: false },
    awardOperationId: { type: Schema.Types.ObjectId, select: false },
    awardWinPoints: { type: Number, min: 0, select: false }
  },
  {
    timestamps: true,
    toJSON: {
      transform(_doc, ret) {
        delete (ret as { __v?: number }).__v;
        delete (ret as Partial<TournamentAttrs>).awardApplied;
        delete (ret as Partial<TournamentAttrs>).awardOperationId;
        delete (ret as Partial<TournamentAttrs>).awardWinPoints;
        return ret;
      }
    }
  }
);

tournamentSchema.index({ status: 1, startDate: 1 });
tournamentSchema.index({ visibility: 1, startDate: -1 });
tournamentSchema.index({ owner: 1, startDate: -1 });
tournamentSchema.index({ admins: 1, startDate: -1 });
tournamentSchema.index({ participants: 1, startDate: -1 });

export const Tournament = model<TournamentAttrs>("Tournament", tournamentSchema);
